src/java/main/org/apache/zookeeper/server/quorum/QuorumCnxManager.java

```diff
@@ -647,11 +648,11 @@ public void run() {
                         numRetries = 0;
                     }
                 } catch (IOException e) {
-                    if (shutdown) {
-                        break;
-                    }
                     LOG.error("Exception while listening", e);
-                    numRetries++;
+                    if (!(e instanceof SocketTimeoutException)) {
+                        numRetries++;
+                    }
+                }finally {
```

Comment:
1. i had removed **finally** code block in my newest code.
2. according to jira , the **SocketTimeoutException** was likely thrown every about 49days, maybe unsigned int -1. Or, we can follow [Karaf](https://issues.apache.org/jira/browse/KARAF-3325l) to fix this code. but i think it was thrown too often.